### PR TITLE
Http client refactor

### LINKED ISSUE #101: "Too many open files" exception
## Observed behavior

This exception causes the middleware to stop responding to requests (see trace below).

Related reading:
- https://stackoverflow.com/questions/55271192/connections-leaking-with-state-close-wait-with-httpclient
- https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8221395

First error - reported in the `*.jar-err.log` file:
```log
Exception in thread "Thread-16062" java.lang.InternalError: java.io.IOException: Too many open files
	at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:311)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.create(HttpClientImpl.java:253)
	at java.net.http/jdk.internal.net.http.HttpClientBuilderImpl.build(HttpClientBuilderImpl.java:135)
	at java.net.http/java.net.http.HttpClient.newHttpClient(HttpClient.java:158)
	at org.opentripplanner.middleware.otp.OtpDispatcher.sendOtpRequest(OtpDispatcher.java:80)
	at org.opentripplanner.middleware.otp.OtpDispatcher.sendOtpPlanRequest(OtpDispatcher.java:51)
	at org.opentripplanner.middleware.tripMonitor.jobs.CheckMonitoredTrip.calculateOtpResponse(CheckMonitoredTrip.java:443)
	at org.opentripplanner.middleware.tripMonitor.jobs.CheckMonitoredTrip.calculateNextItinerary(CheckMonitoredTrip.java:419)
	at org.opentripplanner.middleware.tripMonitor.jobs.CheckMonitoredTrip.shouldSkipMonitoredTripCheck(CheckMonitoredTrip.java:327)
	at org.opentripplanner.middleware.tripMonitor.jobs.CheckMonitoredTrip.doRun(CheckMonitoredTrip.java:102)
	at org.opentripplanner.middleware.tripMonitor.jobs.CheckMonitoredTrip.run(CheckMonitoredTrip.java:92)
	at org.opentripplanner.middleware.tripMonitor.jobs.TripAnalyzer.run(TripAnalyzer.java:72)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: Too many open files
	at java.base/sun.nio.ch.EPoll.create(Native Method)
	at java.base/sun.nio.ch.EPollSelectorImpl.<init>(EPollSelectorImpl.java:79)
	at java.base/sun.nio.ch.EPollSelectorProvider.openSelector(EPollSelectorProvider.java:36)
	at java.base/java.nio.channels.Selector.open(Selector.java:295)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.<init>(HttpClientImpl.java:699)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:308)
	... 12 more
```

Other places involved (from the `*.jar-out.log` file:
```log
java.lang.InternalError: java.io.IOException: Too many open files
	at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:311)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.create(HttpClientImpl.java:253)
	at java.net.http/jdk.internal.net.http.HttpClientBuilderImpl.build(HttpClientBuilderImpl.java:135)
	at java.net.http/java.net.http.HttpClient.newHttpClient(HttpClient.java:158)
	at org.opentripplanner.middleware.otp.OtpDispatcher.sendOtpRequest(OtpDispatcher.java:80)
	at org.opentripplanner.middleware.otp.OtpDispatcher.sendOtpRequest(OtpDispatcher.java:43)
	at org.opentripplanner.middleware.controllers.api.OtpRequestProcessor.proxy(OtpRequestProcessor.java:87)
	at spark.RouteImpl$1.handle(RouteImpl.java:72)
	at spark.http.matching.Routes.execute(Routes.java:61)
	at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)
	at spark.embeddedserver.jetty.JettyHandler.doHandle(JettyHandler.java:50)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1568)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:530)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: Too many open files
	at java.base/sun.nio.ch.IOUtil.makePipe(Native Method)
	at java.base/sun.nio.ch.EPollSelectorImpl.<init>(EPollSelectorImpl.java:83)
	at java.base/sun.nio.ch.EPollSelectorProvider.openSelector(EPollSelectorProvider.java:36)
	at java.base/java.nio.channels.Selector.open(Selector.java:295)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.<init>(HttpClientImpl.java:699)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:308)

```

```log
[31mWARN [0;39m 2020-11-09T23:18:38.238 [33m[qtp1089503494-16-acceptor-0@264014d1-ServerConnector@2887ee69{HTTP/1.1,[http/1.1]}{0.0.0.0:4567}][0;39m [34mo.e.jetty.server.AbstractConnector(AbstractConnector.java:573)[0;39m [33m[0;39m 
java.io.IOException: Too many open files
	at java.base/sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:533)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:285)
	at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:379)
	at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:638)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

```log
Error requesting data from URI
java.io.IOException: java.net.SocketException: Too many open files
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:567)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:119)
	at org.opentripplanner.middleware.utils.HttpUtils.httpRequestRawResponse(HttpUtils.java:99)
	at org.opentripplanner.middleware.utils.HttpUtils.httpRequest(HttpUtils.java:60)
	at org.opentripplanner.middleware.bugsnag.BugsnagDispatcher.makeEventDataRequest(BugsnagDispatcher.java:135)
	at org.opentripplanner.middleware.bugsnag.jobs.BugsnagEventRequestJob.run(BugsnagEventRequestJob.java:29)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.InternalError: java.net.SocketException: Too many open files
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.<init>(PlainHttpConnection.java:224)
	at java.net.http/jdk.internal.net.http.AsyncSSLConnection.<init>(AsyncSSLConnection.java:49)
	at java.net.http/jdk.internal.net.http.HttpConnection.getSSLConnection(HttpConnection.java:239)
	at java.net.http/jdk.internal.net.http.HttpConnection.getConnection(HttpConnection.java:225)
	at java.net.http/jdk.internal.net.http.Http2Connection.createAsync(Http2Connection.java:360)
	at java.net.http/jdk.internal.net.http.Http2ClientImpl.getConnectionFor(Http2ClientImpl.java:127)
	at java.net.http/jdk.internal.net.http.ExchangeImpl.get(ExchangeImpl.java:89)
	at java.net.http/jdk.internal.net.http.Exchange.establishExchange(Exchange.java:299)
	at java.net.http/jdk.internal.net.http.Exchange.responseAsyncImpl0(Exchange.java:431)
	at java.net.http/jdk.internal.net.http.Exchange.responseAsyncImpl(Exchange.java:336)
	at java.net.http/jdk.internal.net.http.Exchange.responseAsync(Exchange.java:328)
	at java.net.http/jdk.internal.net.http.MultiExchange.responseAsyncImpl(MultiExchange.java:346)
	at java.net.http/jdk.internal.net.http.MultiExchange.lambda$responseAsync0$2(MultiExchange.java:292)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:153)
	at java.base/java.util.concurrent.CompletableFuture.completeAsync(CompletableFuture.java:2591)
	at java.net.http/jdk.internal.net.http.MultiExchange.responseAsync(MultiExchange.java:246)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.sendAsync(HttpClientImpl.java:632)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:540)
	... 11 common frames omitted
Caused by: java.net.SocketException: Too many open files
	at java.base/sun.nio.ch.Net.socket0(Native Method)
	at java.base/sun.nio.ch.Net.socket(Net.java:433)
	at java.base/sun.nio.ch.Net.socket(Net.java:426)
	at java.base/sun.nio.ch.SocketChannelImpl.<init>(SocketChannelImpl.java:121)
	at java.base/sun.nio.ch.SelectorProviderImpl.openSocketChannel(SelectorProviderImpl.java:60)
	at java.base/java.nio.channels.SocketChannel.open(SocketChannel.java:150)
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.<init>(PlainHttpConnection.java:213)
	... 31 common frames omitted
```

This Mongo exception may be related:
```log
[34mINFO [0;39m 2020-11-09T00:14:21.411 [33m[cluster-ClusterId{value='5fa02c8964630a086f258ff9', description='null'}-commtrans-otp-shard-00-00.5yaey.mongodb.net:27017][0;39m [34morg.mongodb.driver.cluster(SLF4JLogger.java:76)[0;39m [33m[0;39m Exception in monitor thread while connecting to server commtrans-otp-shard-00-00.5yaey.mongodb.net:27017
com.mongodb.MongoSocketReadException: Prematurely reached end of stream
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:112)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:580)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessage(InternalStreamConnection.java:445)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:299)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:259)
	at com.mongodb.internal.connection.CommandHelper.sendAndReceive(CommandHelper.java:83)
	at com.mongodb.internal.connection.CommandHelper.executeCommand(CommandHelper.java:33)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.initializeConnectionDescription(InternalStreamConnectionInitializer.java:105)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.initialize(InternalStreamConnectionInitializer.java:62)
	at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:129)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:131)
	at java.base/java.lang.Thread.run(Thread.java:834)


```

## Expected behavior

The exception should not occur.

## Steps to reproduce the problem

Two possible ways to reproduce:
1. Save and monitor a number of trips, with SMS trip notifications enabled over 8-10 days.
2. Using batch routing in OTP-RR and using the middleware as a proxy for OTP, run a number of trip plans from the use interface. 

(Actual time to obtain the exception varies on request volume.)

## Any special notes on configuration used

No special configuration, but having a few trips being monitored seems an important factor in this issue.

## Version of otp-middleware and/or OpenTripPlanner if applicable (exact commit hash or branch name)

Use the dev branch.

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR fixes #101 - "Too many files open" exception experienced on the Commtrans instance. The update moves to a single use of HttpClient within the HttpUtils class. The HttpClient reference is now reused for all connections. This is instead of creating new for each connection required.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/utils/HttpUtils.java [hunk 1]
```diff
@@ -70,7 +77,6 @@ public static String httpRequest(URI uri, int connectionTimeout, HttpMethod meth
      */
     public static HttpResponse<String> httpRequestRawResponse(URI uri, int connectionTimeout, HttpMethod method,
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/utils/HttpUtils.java]: I did some searching in the code and found one another (pretty insignificant) use of HttpClient in `PublicApiDocGenerator.class`. I think we should replace that use of `client` in `getSwaggerDocs` with:

```java
return HttpUtils.httpRequestRawResponse(
    URI.create(AUTOGENERATED_DOC_URL), 
    10, 
    HttpMethod.GET, 
    null, 
    null
);
```

This is mainly just to reinforce the use of this util file throughout the app.

Also, one other minor thing: could you change the `connectionTimeout` param to `timeoutInSeconds`?

**Review by landonreed — CHANGES_REQUESTED** [review]
Looks good! A couple of very minor things noted in the comments.

**Review by binh-dam-ibigroup — APPROVED** [review]
What @landonreed said.

File: src/main/java/org/opentripplanner/middleware/utils/HttpUtils.java (hunk 1 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/utils/HttpUtils.java]: Updated.

**Review by landonreed — APPROVED** [review]

**Pull request merged**